lib/puma/reactor.rb

```diff
@@ -122,36 +129,48 @@ def initialize(server, app_pool)
     # This calculation happens in `calculate_sleep`.
     def run_internal
       sockets = @sockets
+      selector = @selector
 
       while true
         begin
-          ready = IO.select sockets, nil, nil, @sleep_for
+          ready = selector.select @sleep_for
         rescue IOError => e
           Thread.current.purge_interrupt_queue if Thread.current.respond_to? :purge_interrupt_queue
-          if sockets.any? { |socket| socket.closed? }
+          if sockets.any? { |socket| m.value.closed? }
             STDERR.puts "Error in select: #{e.message} (#{e.class})"
             STDERR.puts e.backtrace
-            sockets = sockets.reject { |socket| socket.closed? }
+            sockets = sockets.reject do |socket|
+              if m.value.closed?
+                selector.deregister(socket)
+                true
+              end
+            end
+
             retry
           else
             raise
           end
         end
 
-        if ready and reads = ready[0]
-          reads.each do |c|
-            if c == @ready
+        if ready
+          ready.each do |m|
```

Comment:
maybe we can think of something more descriptive than `m`

Reply:
Yup, done.